daemon/cmd/fqdn.go

```diff
@@ -456,10 +461,15 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
 
 	// Ensure that there are no early returns from this function before the
 	// code below, otherwise the log record will not be made.
+	//
+	// Restrict label enrichment time to 10ms; we don't want to block DNS
+	// requests because an identity isn't in the local cache yet.
+	logContext, lcncl := context.WithTimeout(d.ctx, 10*time.Millisecond)
```

Comment:
Given how much lookups are done by `FillEndpointInfo` (called twice by `Addressing`) I'm unsure whether 10ms is a good compromise here.

Reply:
What's your thinking here? Too high, or too low? Remember, we have a latency commitment for DNS requests of 100ms, so 10ms isn't totally out of order here.

The context / timeout is only relevant when fetching identity labels from the kvstore. I'm fairly certain this will never actually happen, since the identities should always be in the local cache.

Specifically, this is a flow we intercepted as a result of L7 policy. With the exception of wildcard selectors, these identities must, by definition, already be in cache, since they're in the policy map. One side of the flow is a local pod, so we must have its identity for policy. The other side of the flow is generally the `kube-dns` pod, which should have a stable identity.

So, this timeout is probably irrelevant, just a circuit-breaker to try and reduce dns proxy tail latency.

Reply:
> What's your thinking here? Too high, or too low?

I had in mind the status-quo of no timeout at all, to which 10ms "seems small" and a potential source of flake.

> Remember, we have a latency commitment for DNS requests of 100ms, so 10ms isn't totally out of order here.

Thanks for pointing it out I wasn't aware of that! Sure 10ms sounds reasonable then.

Now, looking down the call stack the log is sent to the monitor agent, which notify/send its observer/listeners. Concretely, only Hubble subscribe as an observer and only the monitor API register as listener (both enabled by default, controlled by `--enable-hubble` and `--enable-monitor`, respectively).

When both are disabled then we're needlessly paying the logging cost in the hot codepath, and for `--enable-monitor` we're synchronously [gob encoding the message](https://github.com/cilium/cilium/blob/acd761befdbd2d13a4e5dc28887e7d887d4b3afc/pkg/monitor/agent/agent.go#L166-L172) even when there is no active listener at the monitor API level (i.e. no ongoing `cilium monitor`).

If correct then there is room for improvement. I feel like logging, encoding, and metrics accounting should be done asynchronously with respect to the hot DNS request codepath. However that goes beyond the scope PR, should we create a new issue about this?

Reply:
You're absolutely right, we should move this work out of the critical section, and avoid doing it needlessly where possible. That is a bigger change than I'd like for this PR, though. I know @marseel was looking in to deferring the log enrichment.

As an interesting point, we don't want to defer enrichment for too long, as identities and IPs could theoretically be reused. I'm not sure if that's immediately relevant, but it is a consideration.

Reply:
This seems like a fair compromise.

However one concern might be that in the kvstore implementation, under some scenarios, we might be placing some extra load on the kvstore server by starting then cancelling a bunch of requests (although, practically, this shouldn't happen because this will eventually find it's way into the local store?).

Reply:
Do we even coalesce kvstore requests? If not, then we're sending, basically, the same flood of requests.